File: src/main/java/org/timmc/johnny/Ascii7Oracle.java
package org.timmc.johnny;

/**
 * Character set that is limited to 7-bit ASCII. Immutable.
 *
 * Implementation note: This class exists to bridge ImmutableBitSet and
 * CharOracle while deferring the question of how to handle Unicode
 * surrogate pairs. (FIXME)
 */
public class Ascii7Oracle implements CharOracle {

    /** The empty set, for convenience (and preferable to {@link #Ascii7Oracle()}.) */
    public static final Ascii7Oracle EMPTY = new Ascii7Oracle();

    /**
     * Actual backing bitset, where the index of each set bit corresponds to
     * a Java char included in the char set.
     */
    public final ImmutableBitSet backing;

    /*== Constructors ==*/

    /**
     * Create an empty set. Prefer {@link Ascii7Oracle#EMPTY}.
     */
    public Ascii7Oracle() {
        this(ImmutableBitSet.EMPTY);
    }

    /**
     * Create a char set including the indexes of all the set bits in the input.
     * @throws IllegalArgumentException if any characters are outside of 7-bit ASCII
     */
    public Ascii7Oracle(ImmutableBitSet included) {
        validateBitSet(included);
        this.backing = included;
    }

    private static void validateBitSet(ImmutableBitSet bs) {
        int bad = bs.nextSetBit(0x80);
        if (bad != -1) {
            throw new IllegalArgumentException(
                    String.format("Non-lower-ASCII character provided: U+%06X", bad));
        }
    }

    /*== Specified by interfaces ==*/

    public boolean containsChar(char c) {
        return backing.get((int) c);
    }

    public boolean containsCodePoint(int cp) {
        return backing.get(cp);
    }
}

File: src/main/java/org/timmc/johnny/internal/Ascii7Oracle.kt
package org.timmc.johnny.internal

/**
 * Character set that is limited to 7-bit ASCII. Immutable.
 *
 * Implementation note: This class exists to bridge ImmutableBitSet and
 * CharOracle while deferring the question of how to handle Unicode
 * surrogate pairs. (FIXME)
 */
class Ascii7Oracle
/**
 * Create a char set including the indexes of all the set bits in the input.
 * @throws IllegalArgumentException if any characters are outside of 7-bit ASCII
 */
constructor (
    /**
     * Actual backing bitset, where the index of each set bit corresponds to
     * a Java char included in the char set.
     */
    val backing: ImmutableBitSet
) : CharOracle {
    init {
        validateBitSet(backing)
    }

    /*== Specified by interfaces ==*/

    override fun containsChar(c: Char): Boolean {
        return backing.get(c.toInt())
    }

    override fun containsCodePoint(cp: Int): Boolean {
        return backing.get(cp)
    }

    companion object {
        /** The empty set, for convenience.  */
        @JvmField
        val EMPTY = Ascii7Oracle(ImmutableBitSet.EMPTY)

        private fun validateBitSet(bs: ImmutableBitSet) {
            val bad = bs.nextSetBit(0x80)
            if (bad != -1) {
                throw IllegalArgumentException(
                    String.format("Non-lower-ASCII character provided: U+%06X", bad)
                )
            }
        }
    }
}
